### PR TITLE
DCOS-37617 Move menu actions to Graphql

### PR DESCRIPTION
**Pull request by Fabs**
⚠DEPENDS: ~#3040~, ~#3042~

Ignore the merge commits, they will be dropped once we have the dependency PRs in.

**SCOPE**
- Move the delete action to graphql
- Move the stop action to graphql
- Move the toggle schedule action to graphql

**HOW TO TEST**

Basically, go over all the menu scenarios. Stop, Start, Schedule, Remove, Edit. Think a bit of the corner cases too.

**WHAT WAS TAKEN OUT OF SCOPE**

Nothing!

### CODE REVIEW

File: plugins/jobs/src/js/JobDelete.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+import "rxjs/add/operator/combineLatest";
+import "rxjs/add/operator/startWith";
+import "rxjs/add/operator/catch";
+
+import gql from "graphql-tag";
+
+import JobDeleteModal from "./components/JobDeleteModal";
+import defaultSchema from "./data/JobModel";
+
+const deleteJob = gql`
```
Comment (DanielMSchmidt) [inline plugins/jobs/src/js/JobDelete.js]: 🎨 might be only me, but I find the naming confusing. I went ahead and looked up the definition of `graphqlObservable` because I thought a function would be passed to it

File: plugins/jobs/src/js/JobDelete.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+  return {
+    delete$,
+    deleteHandler: (jobId, stopCurrentJobRuns, onSuccess, errorMsg) => {
+      deleteSubject$.next({ jobId, stopCurrentJobRuns, onSuccess, errorMsg });
+    }
+  };
+}
+
+function errorIsTaskCurrentRunning(errorMsg) {
```
Comment (DanielMSchmidt) [inline plugins/jobs/src/js/JobDelete.js]: I like the verbosity of this funcion name a lot :)

File: plugins/jobs/src/js/JobDelete.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+  return /stopCurrentJobRuns=true/.test(errorMsg);
+}
+
+const JobDelete = componentFromStream(prop$ => {
+  const { delete$, deleteHandler } = deleteOperation();
```
Comment (DanielMSchmidt) [inline plugins/jobs/src/js/JobDelete.js]: We would create a new subject every time the props change, does this really make sense?
(I see this in multiple occasions but choose to only comment it here to discuss it more easily)

File: plugins/jobs/src/js/JobDelete.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  const deleteEmit$ = delete$.startWith({ done: null });
+
+  return prop$
+    .combineLatest(deleteEmit$, (props, deleteOp) => {
+      return { ...props, ...deleteOp };
+    })
+    .map(({ open, jobId, onClose, onSuccess, errorMsg, done }) => {
+      const isDisabled = done === false;
+      const isOpen = open;
```
Comment (DanielMSchmidt) [inline plugins/jobs/src/js/JobDelete.js]: This variable could easily be removed :)

File: plugins/jobs/src/js/JobDelete.js [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+const JobDelete = componentFromStream(prop$ => {
+  const { delete$, deleteHandler } = deleteOperation();
+  const deleteEmit$ = delete$.startWith({ done: null });
+
+  return prop$
+    .combineLatest(deleteEmit$, (props, deleteOp) => {
+      return { ...props, ...deleteOp };
+    })
+    .map(({ open, jobId, onClose, onSuccess, errorMsg, done }) => {
+      const isDisabled = done === false;
```
Comment (DanielMSchmidt) [inline plugins/jobs/src/js/JobDelete.js]: What would you think of inlining this? My reasoning is that if you see this at the top of this function you need to keep it in mind and see if anything reacts to it, just to see that it is simply passed through

File: plugins/jobs/src/js/JobDelete.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+      const isOpen = open;
+
+      const stopCurrentJobRuns = errorIsTaskCurrentRunning(errorMsg);
+
+      const onSuccessEvent = () => {
+        deleteHandler(jobId, stopCurrentJobRuns, onSuccess, errorMsg);
+      };
+
+      const onCloseEvent = () => {
+        onClose();
```
Comment (DanielMSchmidt) [inline plugins/jobs/src/js/JobDelete.js]: Can't we simply pass onClose into the modal?

File: plugins/jobs/src/js/JobStopRun.js [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+  return props$
+    .combineLatest(stopEmit$, (props, stopOperation) => {
+      return { ...props, ...stopOperation };
+    })
+    .map(({ jobID, jobRuns, onClose, onSuccess, open, done }) => {
+      const disabled = done === false;
+
+      const onSuccessEvent = () => {
+        if (jobRuns.length === 1) {
```
Comment (DanielMSchmidt) [inline plugins/jobs/src/js/JobStopRun.js]: This is really confusing to me, why do we check for `jobRuns`?

File: plugins/jobs/src/js/JobStopRun.js [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+  return {
+    stop$,
+    stopHandler: (jobId, jobRunId, onSuccess) => {
+      stopSubject$.next({ jobId, jobRunId, onSuccess });
+    }
+  };
+}
+
+const JobStopRun = componentFromStream(props$ => {
```
Comment (DanielMSchmidt) [inline plugins/jobs/src/js/JobStopRun.js]: Can we please add some basic tests for this? I would at least like to test the `if (jobRuns.length === 1) {` line below, see if a mutation goes through and if we can deal with an API failure gracefully

File: plugins/jobs/src/js/components/JobDeleteModal.js [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+  };
+}
+
+function modalContent(jobId, stopCurrentJobRuns) {
+  const contentBuilder = stopCurrentJobRuns ? stopJobsCotent : defaultContent;
+
+  return contentBuilder(jobId);
+}
+
+function modalContentHtml(message) {
```
Comment (DanielMSchmidt) [inline plugins/jobs/src/js/components/JobDeleteModal.js]: Let's make this a component on its own 👍

File: plugins/jobs/src/js/components/JobDeleteModal.js [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+
+  const defaultLabel = `${StringUtil.capitalize(UserActions.DELETE)} Job`;
+
+  return {
+    label: defaultLabel,
+    message: modalContentHtml(defaultMessage)
+  };
+}
+
+function stopJobsCotent(jobId) {
```
Comment (DanielMSchmidt) [inline plugins/jobs/src/js/components/JobDeleteModal.js]:
```diff
- function stopJobsCotent(jobId) {
+ function stopJobsContent(jobId) {
```

File: plugins/jobs/src/js/components/JobDeleteModal.js [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    UserActions.DELETE
+  } the job?`;
+
+  return {
+    label: stopCurrentRunsLabel,
+    message: modalContentHtml(stopCurrentRunsMessage)
+  };
+}
+
+function modalContent(jobId, stopCurrentJobRuns) {
```
Comment (DanielMSchmidt) [inline plugins/jobs/src/js/components/JobDeleteModal.js]: This is one of the rare cases where I think that having these put into functions makes it more unreadable and harder to understand. Can we either do an if on the call-side or get rid of this builder pattern?

File: plugins/jobs/src/js/components/JobStopRunModal.js [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+        onClose={onClose}
+        leftButtonText="Cancel"
+        leftButtonCallback={onClose}
+        leftButtonClassName="button button-primary-link"
+        rightButtonText={rightButtonText}
+        rightButtonClassName="button button-danger"
+        rightButtonCallback={this.props.onSuccess}
+        showHeader={true}
+      >
+        {this.getModalContents()}
```
Comment (DanielMSchmidt) [inline plugins/jobs/src/js/components/JobStopRunModal.js]: 🎨 we could also just inline this

File: plugins/jobs/src/js/data/JobModel.ts [hunk 13]
```diff
@@ -141,7 +156,31 @@ export const resolvers = ({
       args: GeneralArgs,
       _context = {}
     ): Observable<JobLink> {
-      return runJob(args.id).map(({ jobId }) => ({ jobId }));
+      if (args.id) {
```
Comment (DanielMSchmidt) [inline plugins/jobs/src/js/data/JobModel.ts]: What would you think about using a special interface here and using a user-defined type check?

File: plugins/jobs/src/js/data/JobModel.ts [hunk 14]
```diff
@@ -141,7 +156,31 @@ export const resolvers = ({
...
+        return runJob(args.id).map(({ jobId }) => ({ jobId }));
+      }
+
+      return Observable.throw({
```
Comment (DanielMSchmidt) [inline plugins/jobs/src/js/data/JobModel.ts]: I think we need to talk about what we want to throw as a team. I in the other resolvers we just throw strings and I would like to see it concise, but I think having an object with additional info is also nice 🤔

File: plugins/jobs/src/js/JobDelete.js (hunk 3 above)
Comment (Fabs) [reply plugins/jobs/src/js/JobDelete.js]: I can verify that. But I think that is not the case. componentFromStream will be called only once to generate the component. I believe this function will be called on construction or mounting. 

Notice that this is outside of the `prop$.combine.map..` which means that it does not get called on every property refresh.

File: plugins/jobs/src/js/JobDelete.js (hunk 4 above)
Comment (Fabs) [reply plugins/jobs/src/js/JobDelete.js]: Sure, Thanks.

File: plugins/jobs/src/js/JobDelete.js (hunk 6 above)
Comment (Fabs) [reply plugins/jobs/src/js/JobDelete.js]: Yeap.

File: plugins/jobs/src/js/JobStopRun.js (hunk 7 above)
Comment (Fabs) [reply plugins/jobs/src/js/JobStopRun.js]: This is a legacy implementation, I will restore its comment.

You have a list where you can select many jobs, but the list only allows you to select one (because the implementation is not complete), so this bit test that your expectations are met, and you can only stop one job at a time. (although the nice Ux would be allowing it to do many).

File: plugins/jobs/src/js/JobStopRun.js (hunk 8 above)
Comment (Fabs) [reply plugins/jobs/src/js/JobStopRun.js]: I agree.

File: plugins/jobs/src/js/JobDelete.js (hunk 3 above)
Comment (DanielMSchmidt) [reply plugins/jobs/src/js/JobDelete.js]: Oh okay, please verify. If it is like you say then it's totally fine

File: plugins/jobs/src/js/data/JobModel.ts (hunk 14 above)
Comment (Fabs) [reply plugins/jobs/src/js/data/JobModel.ts]: I agree, and I would love to discuss that here. But changing it I believe would be out of the scope since I am following the API all the other resolvers use, and I think we should fix them all together. 

I think it is maybe a topic for the patterns we wanna have with the data layer.

File: plugins/jobs/src/js/components/JobDeleteModal.js (hunk 11 above)
Comment (Fabs) [reply plugins/jobs/src/js/components/JobDeleteModal.js]: I agree with you here and the HTML function. I did the `quick extraction ™`

File: plugins/jobs/src/js/components/JobStopRunModal.js (hunk 12 above)
Comment (Fabs) [reply plugins/jobs/src/js/components/JobStopRunModal.js]: Yes, now that is short enough that we can.

File: plugins/jobs/src/js/data/JobModel.ts (hunk 13 above)
Comment (Fabs) [reply plugins/jobs/src/js/data/JobModel.ts]: I like that idea. But I think it is adding one more level of complexity to something that should not exist. Graphql should check the existence and type of my args, I should not have to add all those guards to the resolver.

I think here I would also make "the same mistake" everywhere and push forward to have our graphql implementation to care about it.

File: plugins/jobs/src/js/JobDelete.js (hunk 5 above)
Comment (Fabs) [reply plugins/jobs/src/js/JobDelete.js]: fair point.

File: plugins/jobs/src/js/data/JobModel.ts (hunk 14 above)
Comment (DanielMSchmidt) [reply plugins/jobs/src/js/data/JobModel.ts]: I think we missed that already and we are unaligned already: https://github.com/dcos/dcos-ui/blob/8429c574a2a99c7704127ee4029a90f69302df55/plugins/jobs/src/js/data/JobModel.ts#L99

File: plugins/jobs/src/js/data/JobModel.ts (hunk 13 above)
Comment (DanielMSchmidt) [reply plugins/jobs/src/js/data/JobModel.ts]: Can you explain what you mean with "our graphql implementation to care about it"

Comment (Fabs) [reply plugins/jobs/src/js/data/JobModel.ts]: For Runtime:
The query this resolver resolves states: `runJob(id: String!)`. Which means that the grapqhl implementation should throw an error if `id` is not present. So at runtime, I expect not to have to write an if.

For type checking:
As for the typescript types, I understand and like the elegance of the user-defined type check like `isJobQueryArg()`, but on this particular case I believe it just hides a cast away from me.  

What I really want to right is:
`args: RunJobArgs`

but then IResolvers do not get happy. That I think is the type who needs a tweek to be able to support me to say `args: RunJobArgs`.

So overall I think the `if` here is also the simple solution to those both problems combined.

File: plugins/jobs/src/js/JobDelete.js (hunk 2 above)
Comment (nLight) [reply plugins/jobs/src/js/JobDelete.js]: It's hard to understand what's going on from the name. Will it error out? Will it detect error? Shall it be `errorIfTaskCurrentRunning` or `errorIfCurrentTaskIsRunning`?

File: plugins/jobs/src/js/data/JobModel.ts (hunk 14 above)
Comment (Fabs) [reply plugins/jobs/src/js/data/JobModel.ts]: I have no real preference between the `String` and the `response.mesage.String`. if you have a preference I would use it.

File: plugins/jobs/src/js/data/JobModel.ts (hunk 13 above)
Comment (Fabs) [reply plugins/jobs/src/js/data/JobModel.ts]: does that argument, solve this issue?

File: plugins/jobs/src/js/JobDelete.js (hunk 2 above)
Comment (Fabs) [reply plugins/jobs/src/js/JobDelete.js]: What about: isTaskCurrentRunning?

File: plugins/jobs/src/js/JobDelete.js (hunk 3 above)
Comment (Fabs) [reply plugins/jobs/src/js/JobDelete.js]: It is only called once.

**Comment by Fabs** [comment]
@nLight and @DanielMSchmidt . I fixed it, but I am not setting it to review again until the other dependencies are in.

File: plugins/jobs/src/js/JobDelete.js (hunk 2 above)
Comment (nLight) [reply plugins/jobs/src/js/JobDelete.js]: Is `TaskCurrent` an entity? Why then not `CurrentTask`?

File: plugins/jobs/src/js/JobsOverview.tsx [hunk 15]
```diff
@@ -158,7 +158,7 @@ const JobsOverview = withRouter(
           return graphqlObservable(jobsOverviewQuery, JobModel, {
             path,
             filter
-          }) as Observable<{ data: { jobs: JobConnection } }>;
```
Comment (Fabs) [inline plugins/jobs/src/js/JobsOverview.tsx]: Should vanish once we get this fix on another PR to master.

**Comment by nLight** [comment]
Flicts!

**Comment by nLight** [comment]
All deps are in, please resolve the flicts!

File: plugins/jobs/src/js/JobDelete.js [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+        jobId,
+        stopCurrentJobRuns,
+        onSuccess,
+        errorMessage
+      });
+    }
+  };
+}
+
+function isTaskCurrentRunning(errorMessage) {
```
Comment (nLight) [inline plugins/jobs/src/js/JobDelete.js]: I hate to get back to this but Task preceding Current still doesn't make any sense to me. Am I missing something?

Comment (Fabs) [reply plugins/jobs/src/js/JobDelete.js]: I think what is wrong here is that it should be written "currently". Does that work for you? `isTaskCurrentlyRunning`

Comment (nLight) [reply plugins/jobs/src/js/JobDelete.js]: that would make total sense 🙇

File: plugins/jobs/src/js/JobDelete.js [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+  return /stopCurrentJobRuns=true/.test(errorMessage);
```
[the remark this answers is not in the document]
Comment (juliangieseke) [reply plugins/jobs/src/js/JobDelete.js]: haha, was going to comment, but thoght its not important enough :D

**Comment by Fabs** [comment]
That PR is dead, CI wont build it anymore :/

**Comment by Fabs** [comment]
I moved to a new PR this seems not to build.
